### DISCUSSION TITLE
Can I use hardhat as a JS library without polluting the global scope?

### DISCUSSION THREAD

**Discussion by evbots** [NomicFoundation/hardhat#3804] [General]
Hi hardhat community,

I'm attempting to use hardhat as a javascript library inside a web server to compile contracts "on the fly".

My web server does these things:

1. saves a .sol file based on user input
2. compiles the contract to create artifacts
3. returns bytecode and ABI in the response

Each user request requires saving contract output in a different artifacts directory, which is why I need to pass in that config into an Environment constructor (as shown below in my example).

I am using hardhat successfully, but I've noticed that I'm not able to just create objects in local scope. I can't seem to create an `Environment` object without creating a `HardhatContext` object, which appears to inject a variable into the global scope.

Is there a way to use hardhat that does not inject objects into the global scope? This would address my worries that there is some kind of race condition happening inside global scope if the web server is serving multiple requests at the same time.

Thank you! Example below.

```javascript
const { Environment } = require("hardhat/internal/core/runtime-environment");
const { HardhatContext } = require("hardhat/internal/context");
const { getEnvHardhatArguments } = require("hardhat/internal/core/params/env-variables");
const { HARDHAT_PARAM_DEFINITIONS } = require("hardhat/internal/core/params/hardhat-params");
const { loadConfigAndTasks } = require("hardhat/internal/core/config/config-loading");
const { TasksDSL } = require("hardhat/internal/core/tasks/dsl");


let ctx;

const compile = async (directory_folder) => {
  if (HardhatContext.isCreated()) {
    ctx = HardhatContext.getHardhatContext();
  } else {
    ctx = HardhatContext.createHardhatContext(); 
  }

  const hardhatArguments = getEnvHardhatArguments(
    HARDHAT_PARAM_DEFINITIONS,
    process.env
  );

  const { resolvedConfig, userConfig } = loadConfigAndTasks(hardhatArguments);

  resolvedConfig["paths"]["artifacts"] = `${__dirname}/generated_artifacts`;
  resolvedConfig["paths"]["cache"] = `${__dirname}/generated_cache/${directory_folder}`;
  resolvedConfig["paths"]["sources"] = `${__dirname}/generated_contracts/${directory_folder}`;

  const hre = new Environment(
    resolvedConfig,
    hardhatArguments,
    ctx.tasksDSL.getTaskDefinitions(),
    ctx.extendersManager.getExtenders(),
    ctx.experimentalHardhatNetworkMessageTraceHooks,
    userConfig
  );

  await hre.run("compile");  
};
```

**Comment by fvictorio** [comment]
Hi @evbots. This is not possible right now. We want to add some way to programmatically create Hardhat contexts, which wouldn't modify the global scope and would probably be what you want. But we don't have that feature yet.

Without knowing a lot about your app, my guess is that your best shot is to call Hardhat in a sub-process. For example, you can do something like this:

```js
const { exec } = require("child_process")

const child = exec("npx hardhat compile", { cwd: pathToProjectRoot })
// read stdout and stderr from child, wait for it to finish, etc.
```

There's a lot to hate about this solution, and there are several ways it can be improved (like using `util.promisify` to get a better API, or use `fork` instead of `exec` for better performance), but maybe you find it useful. Again, I don't know the details of your use case so this might not be an option for you.

**Reply by evbots** [reply]
using `exec` or `fork` is a great idea. thank you for the suggestion, that should solve the immediate problem.

I would also love to see a programmatic way to create hardhat contexts in the future ❤️
